frontend/apps/reader/modules/readerbookmark.lua

```diff
@@ -385,24 +396,26 @@ function ReaderBookmark:updateHighlightsIfNeeded()
     self.ui.doc_settings:saveSetting("bookmarks_version", 20200615)
 end
 
-function ReaderBookmark:onShowBookmark()
+function ReaderBookmark:onShowBookmark(filter_table)
     self.select_mode = false
-    self.filtered_mode = false
+    self.filtered_mode = filter_table and true or false
     self:updateHighlightsIfNeeded()
     -- build up item_table
     local item_table = {}
     local is_reverse_sorting = G_reader_settings:nilOrTrue("bookmarks_items_reverse_sorting")
     local num = #self.bookmarks + 1
-    for i, v in ipairs(self.bookmarks) do
+    local count_not_found = 0
+    for i = 1, #self.bookmarks do
+        -- bookmarks are internally sorted by descending page numbers
+        local v = self.bookmarks[is_reverse_sorting and i or num - i]
         local is_auto_text
         if v.text == nil or v.text == "" then
             is_auto_text = true
             v.text = self:getBookmarkAutoText(v)
         else
             is_auto_text = self:isBookmarkAutoText(v)
         end
-        -- bookmarks are internally sorted by descending page numbers
-        local k = is_reverse_sorting and i or num - i
+        local k = i - count_not_found
```

Comment:
Not sure I'm right, but there, you get consecutive ascending index `k` (I initially thought it was more complicated than that :). So, if that is right, below you could just use `table.insert(item_table, the_deep_copy_you_made)`.

It feels in Lua `item_table[k] =` might involve more work than `table.insert()` (not sure when k is consecutive ascending, but I'm pretty sure it's worse with `k` in random order.